lightning/src/routing/scoring.rs

```diff
@@ -460,25 +454,423 @@ impl<T: Time> Readable for ChannelFailure<T> {
+/// Then the negative log of the success probability is used to determine the cost of routing a
+/// specific HTLC amount through a channel.
+///
+/// Knowledge about channel liquidity balances takes the form of upper and lower bounds on the
+/// possible liquidity. Certainty of the bounds is decreased over time using a decay function. See
+/// [`ProbabilisticScoringParameters`] for details.
+///
+/// [1]: https://arxiv.org/abs/2107.05322
+pub type ProbabilisticScorer<G> = ProbabilisticScorerUsingTime::<G, ConfiguredTime>;
+
+/// Probabilistic [`Score`] implementation.
+///
+/// See [`ProbabilisticScorer`] for details.
+///
+/// # Note
+///
+/// Mixing the `no-std` feature between serialization and deserialization results in undefined
+/// behavior.
+///
+/// (C-not exported) generally all users should use the [`ProbabilisticScorer`] type alias.
+pub struct ProbabilisticScorerUsingTime<G: Deref<Target = NetworkGraph>, T: Time> {
+	params: ProbabilisticScoringParameters,
+	node_id: NodeId,
+	network_graph: G,
+	// TODO: Remove entries of closed channels.
+	channel_liquidities: HashMap<u64, ChannelLiquidity<T>>,
+}
+
+/// Parameters for configuring [`ProbabilisticScorer`].
+pub struct ProbabilisticScoringParameters {
+	/// The time required to elapse before any knowledge learned about channel liquidity balances is
+	/// cut in half.
+	///
+	/// The bounds are defined in terms offsets and are initially zero. Increasing the offsets gives
+	/// tighter bounds on the channel liquidity balance. Thus, halving the offsets decreases the
+	/// certainty of the channel liquidity balance.
+	///
+	/// Default value: 1 hour
+	///
+	/// # Note
+	///
+	/// When built with the `no-std` feature, time will never elapse. Therefore, the channel
+	/// liquidity knowledge will never decay except when the bounds cross.
+	pub liquidity_offset_half_life: Duration,
+}
+
+impl_writeable_tlv_based!(ProbabilisticScoringParameters, {
+	(0, liquidity_offset_half_life, required),
+});
+
+/// Accounting for channel liquidity balance uncertainty.
+///
+/// Direction is defined in terms of [`NodeId`] partial ordering, where the source node is the
+/// first node in the ordering of the channel's counterparties. Thus, swapping the two liquidity
+/// offset fields gives the opposite direction.
+struct ChannelLiquidity<T: Time> {
+	min_liquidity_offset_msat: u64,
+	max_liquidity_offset_msat: u64,
+	last_updated: T,
+}
+
+/// A view of [`ChannelLiquidity`] in one direction assuming a certain channel capacity and decayed
+/// with a given half life.
+struct DirectedChannelLiquidity<L: Deref<Target = u64>, T: Time, U: Deref<Target = T>> {
+	min_liquidity_offset_msat: L,
+	max_liquidity_offset_msat: L,
+	capacity_msat: u64,
+	last_updated: U,
+	now: T,
+	half_life: Duration,
+}
+
+/// The likelihood of an event occurring.
+enum Probability {
+	Zero,
+	One,
+	Ratio { numerator: u64, denominator: u64 },
```

Comment:
Right, fair, avoiding the FPU in the inner loop of the router may be useful, but given this is an internal API so we can change it later, I don't see why we should eat the complexity now vs just adding it later if we want to do some approximation. Sadly our benchmarks for routing currently don't hit the scoring code much because the scorer is uninitialized initially. Doesn't need to happen in this PR, but ideally we'd add the scorer to the benchmark(s).

Reply:
Leaving in for now to try and update the benchmark in parallel with review, but happy to remove it before merge if needed.

Reply:
So you're saying we should just use a linear division instead of log entirely?